### PR TITLE
[addons] add GetTime... function and possibility to get ID3D11DeviceContext all the time

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description

Commit 1:
-----------------------------------------------
**[addons]** add possibility to get ID3D11DeviceContext all the time

Have seen by help of @afedchin that the context can be change on the fly.
Before was to Screensaver and Visualization this only given during start.

This adds a new addon interface function (`void* kodi::gui::GetHWContext()`)
where the addon can ask about them if needed.

Commit 2:
-----------------------------------------------
**[addons]** this adds helper functions to get current time

The code happens on [p8platform](https://github.com/Pulse-Eight/platform/blob/master/src/util/timeutils.h#L78), but since this defines "long"
as value, it is useless because it is too inaccurate in frames.

Since a lot of addons (over 20) need this I think it belongs to the dev kit.

The code is completely independent of Kodi and gives no access to it.

## Motivation and Context
Improve and make DirectX more safe

## How Has This Been Tested?
On screensaver addons

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by afedchin — CHANGES_REQUESTED** [review]
to avoid a lot of ifdefs let's introduce a method in windowing `GetHWContext` and use it via service broker. on windows windowing has pointer to `DeviceResources`

File: xbmc/windowing/win10/WinSystemWin10.h [hunk 1]
```diff
@@ -101,6 +101,7 @@ class CWinSystemWin10 : public CWinSystemBase
 
 protected:
   bool CreateNewWindow(const std::string& name, bool fullScreen, RESOLUTION_INFO& res) override = 0;
+  void* GetHWContext() override = 0;
```
Comment (afedchin) [inline xbmc/windowing/win10/WinSystemWin10.h]: I see no reason to add this.

File: xbmc/windowing/windows/WinSystemWin32.h [hunk 2]
```diff
@@ -128,6 +128,7 @@ class CWinSystemWin32 : public CWinSystemBase
 
 protected:
   bool CreateNewWindow(const std::string& name, bool fullScreen, RESOLUTION_INFO& res) override = 0;
+  void* GetHWContext() override = 0;
```
Comment (afedchin) [inline xbmc/windowing/windows/WinSystemWin32.h]: same as above

File: xbmc/windowing/win10/WinSystemWin10.h (hunk 1 above)
Comment (AlwinEsch) [reply xbmc/windowing/win10/WinSystemWin10.h]: Seen it on the other and thought to do the same but you are right, by `CreateNewWindow` the CWinSystemBase is also defined with `=0` and then makes it sense.

**Comment by AlwinEsch** [comment]
How is the current merge window, can be this in or wait?

jenkins build this please

**Comment by afedchin** [comment]
wait until Leia will being branched

**Review by afedchin — APPROVED** [review]

**Comment by Rechi** [comment]
wrong revision was build
jenkins build this please

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/General.h [hunk 3]
```diff
@@ -141,5 +141,35 @@ namespace gui
...
+  /// \note Returned Windows class pointer is `ID3D11DeviceContext1`.
+  ///
+  ///
+  ///-------------------------------------------------------------------------
+  ///
+  /// **Example:**
+  /// ~~~~~~~~~~~~~{.cpp}
+  /// #include <d3d11_1.h>
+  /// ..
+  /// ID3D11DeviceContext1* context = reinterpret_cast<ID3D11DeviceContext1>(kodi::gui::GetD3DContext());
```
Comment (garbear) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/General.h]: This line doesn't look right

Comment (AlwinEsch) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/General.h]: Ups, becomes changed to `GetHWContext` (Was the name before)

Comment (afedchin) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/General.h]:
```suggestion
  /// ID3D11DeviceContext1* context = reinterpret_cast<ID3D11DeviceContext1*>(kodi::gui::GetHWContext());
```
wrong reinterpret_cast's type param :)

Comment (garbear) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/General.h]: Also static_cast might suffice?

Comment (AlwinEsch) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/General.h]: Is updated, yes with static_cast also worked and is changed.

**Comment by AlwinEsch** [comment]
jenkins build this please

OK to bring in?

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+#elif defined(TARGET_WINDOWS)
+#include <Windows.h>
+#include <time.h>
+#else
+#include <time.h>
+#endif
+
+namespace kodi
+{
+  namespace time
```
Comment (Rechi) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h]: namespaces are not indented, https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    ///
+    /// ...
+    /// double time = kodi::time::GetTimeMs();
+    /// ...
+    /// ~~~~~~~~~~~~~
+    ///
+    inline double GetTimeMs()
+    {
+    #if defined(TARGET_DARWIN)
+      return (double) (CVGetCurrentHostTime() / (double)(CVGetHostClockFrequency() * 0.001));
```
Comment (Rechi) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h]: C++ style casts

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    #elif defined(TARGET_WINDOWS)
+      LARGE_INTEGER tickPerSecond;
+      LARGE_INTEGER tick;
+      if (QueryPerformanceFrequency(&tickPerSecond))
+      {
+        QueryPerformanceCounter(&tick);
+        return (double) (tick.QuadPart / (tickPerSecond.QuadPart / 1000.));
```
Comment (Rechi) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h]: C++ style cast

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+      }
+      return -1;
+    #else
+      timespec time;
+      clock_gettime(CLOCK_MONOTONIC, &time);
+      return (double)time.tv_sec * 1000.0 + time.tv_nsec / 1000000.0;
```
Comment (Rechi) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h]: C++ style cast

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    ///
+    /// ...
+    /// double time = kodi::time::GetTimeSec<double>();
+    /// ...
+    /// ~~~~~~~~~~~~~
+    ///
+    template <class T>
+    inline T GetTimeSec()
+    {
+      return (T)GetTimeMs() / (T)1000.0;
```
Comment (Rechi) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h]: C++ style casts

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h [hunk 9]
```diff
@@ -44,7 +44,7 @@
 #define ADDON_GLOBAL_VERSION_GENERAL_XML_ID           "kodi.binary.global.general"
 #define ADDON_GLOBAL_VERSION_GENERAL_DEPENDS          "General.h"
 
-#define ADDON_GLOBAL_VERSION_GUI                      "5.12.0"
+#define ADDON_GLOBAL_VERSION_GUI                      "5.12.1"
```
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h]: Is this a compatible API change, means, addons must not be recompiled?

Comment (AlwinEsch) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h]: Yes, it add only the header for time and for the DirectX added function is on a by `malloc` created structure entry on end.  This is then a bit bigger but becomes never edited on addon side.

https://github.com/xbmc/xbmc/pull/15609/files/a51776e52d7301a7de8e39a28b1b7efdffd7b6c1#diff-c244971ddb8feeb77e6e8a8c7af39a99R29

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h (hunk 5 above)
Comment (AlwinEsch) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h]: I look to change them here and on other places, was taken like them from p8platform.

Can someone quikly test that on Mac a:
```cpp
return static_cast<double>(CVGetCurrentHostTime() / static_cast<double>(CVGetHostClockFrequency() * 0.001));
```
works correct.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h (hunk 9 above)
Comment (ksooo) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h]: Key for compatibility is to add the new function table entry at the end of the table, as you did. All fine, then.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h (hunk 5 above)
Comment (AlwinEsch) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/tools/Time.h]: The code is updated and tested on Windows and Linux, only not Mac, think it works but not confirmed!

**Comment by AlwinEsch** [comment]
jenkins build this please

**Pull request merged**